homeassistant/components/xiaomi_aqara/__init__.py

```diff
@@ -201,23 +208,32 @@ def stop_xiaomi(event):
     return True
 
 
-async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
     """Unload a config entry."""
-    if entry.data[CONF_KEY] is not None:
+    if config_entry.data[CONF_KEY] is not None:
         platforms = GATEWAY_PLATFORMS
     else:
         platforms = GATEWAY_PLATFORMS_NO_KEY
 
-    unload_ok = await hass.config_entries.async_unload_platforms(entry, platforms)
+    unload_ok = await hass.config_entries.async_unload_platforms(
+        config_entry, platforms
+    )
     if unload_ok:
-        hass.data[DOMAIN][GATEWAYS_KEY].pop(entry.entry_id)
-
-    if len(hass.data[DOMAIN][GATEWAYS_KEY]) == 0:
+        hass.data[DOMAIN][GATEWAYS_KEY].pop(config_entry.entry_id)
+
+    loaded_entries = [
+        entry
+        for entry in hass.config_entries.async_entries(DOMAIN)
+        if entry.state == ConfigEntryState.LOADED
+    ]
+    if len(loaded_entries) == 1:
         # No gateways left, stop Xiaomi socket
+        unsub_stop = hass.data[DOMAIN].pop(KEY_UNSUB_STOP)
+        unsub_stop()
         hass.data[DOMAIN].pop(GATEWAYS_KEY)
         _LOGGER.debug("Shutting down Xiaomi Gateway Listener")
-        gateway_discovery = hass.data[DOMAIN].pop(LISTENER_KEY)
-        await hass.async_add_executor_job(gateway_discovery.stop_listen)
+        multicast = hass.data[DOMAIN].pop(LISTENER_KEY)
+        await hass.async_add_executor_job(multicast.stop_listen)
```

Comment:
Why is this run in the thread pool?

Reply:
Because I did not know if `self.transport.close()` and `self._sock.close()` were async.
But I now know that they can and schould be run from within the event loop.
Thanks!